### PR TITLE
[ML] fix custom feature processor extraction bugs around boolean fields and custom one_hot feature output order

### PR DESCRIPTION
**Pull request by benwtrent**
This commit fixes two problems:

- When extracting a doc value, we allow boolean scalars to be used as input
- The output order of processed feature names is deterministic. Previous custom one hot fields used to be non-deterministic and thus could cause weird bugs.

### CODE REVIEW

**Comment by benwtrent** [comment]
@elasticmachine update branch

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/extractor/ProcessedField.java [hunk 1]
```diff
@@ -24,20 +27,21 @@ public ProcessedField(PreProcessor processor) {
     }
 
     public List<String> getInputFieldNames() {
-        return preProcessor.inputFields();
+        return preProcessor.inputFields().stream().sorted().collect(Collectors.toList());
     }
 
     public List<String> getOutputFieldNames() {
-        return preProcessor.outputFields();
+        return preProcessor.outputFields().stream().sorted().collect(Collectors.toList());
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/extractor/ProcessedField.java:34]: Ditto

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/extractor/ProcessedField.java [hunk 2]
```diff
@@ -24,20 +27,21 @@ public ProcessedField(PreProcessor processor) {
     }
 
     public List<String> getInputFieldNames() {
-        return preProcessor.inputFields();
+        return preProcessor.inputFields().stream().sorted().collect(Collectors.toList());
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/extractor/ProcessedField.java:30]: This is called for every document we load into the process. I suggest we sort once when we store them in the `PreProcessor`.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/extractor/ProcessedField.java [hunk 3]
```diff
@@ -47,12 +51,12 @@ public ProcessedField(PreProcessor processor) {
...
             }
             final Object value = values[0];
-            if (values.length == 1 && (value instanceof String || value instanceof Number)) {
+            if (values.length == 1 && (isValidValue(value))) {
                 inputs.put(field, value);
             }
         }
         preProcessor.process(inputs);
-        return preProcessor.outputFields().stream().map(inputs::get).toArray();
+        return preProcessor.outputFields().stream().sorted().map(inputs::get).toArray();
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/extractor/ProcessedField.java:59]: If the preprocessor already returns the output fields sorted, we shouldn't need to sort here.

Now unfortunately there is not sorted list interface. But as we're dealing with fields, if we want to enforce this to the PreProcessor interface we should switch to `SortedSet`.

Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/extractor/ProcessedField.java:59]: The only one that causes this problem is the Onehot processor, I will see if I can figure out a way for the order to always be the same. The order will always have to be the same after serializing from the wire and on xcontent.

Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/extractor/ProcessedField.java:59]: Yeah, I don't know what I was thinking 🤦 

I will fix this...The preprocessor output should be consistent

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM

**Pull request merged**